File: LeetCode/cpp/1329_Sort_the_Matrix_Diagonally.cpp
class Solution {
public:
	vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
		map<pair<int, int>, vector<int>> um;

		for (int i = 0; i < mat.size(); i++) {
			for (int j = 0; j < mat[0].size(); j++) {
				if (i <= j) {
					um[{i - i, j - i}].push_back(mat[i][j]);
				}
				else {
					um[{i - j, j - j}].push_back(mat[i][j]);
				}
			}
		}

		for (auto m : um) {
			vector<int> v = m.second;

			sort(v.begin(), v.end());

			int x = m.first.first;
			int y = m.first.second;

			for (int i = 0; i < v.size(); i++) {
				mat[x][y] = v[i];

				x++;
				y++;
			}
		}

		return mat;
	}
};

File: LeetCode/java/1329_Sort_the_Matrix_Diagonally.java
import java.util.*;

class Solution {
    public int[][] diagonalSort(int[][] mat) {
        Map<Pair, List<Integer>> map = new HashMap<>();

        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j < mat[0].length; j++) {
                Pair pair;
                if (i <= j) {
                    pair = new Pair(i - i, j - i);
                } else {
                    pair = new Pair(i - j, j - j);
                }

                if (!map.containsKey(pair)) {
                    map.put(pair, new ArrayList<>());
                }

                map.get(pair).add(mat[i][j]);
            }
        }

        for (Pair key : map.keySet()) {
            List<Integer> list = map.get(key);

            Collections.sort(list);

            int x = key.getX();
            int y = key.getY();

            for (int i = 0; i < list.size(); i++) {
                mat[x][y] = list.get(i);

                x++;
                y++;
            }
        }

        return mat;
    }
}

class Pair {
    int x;
    int y;

    Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }

    int getX() {
        return x;
    }

    int getY() {
        return y;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pair pair = (Pair) o;
        return x == pair.x &&
                y == pair.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }
}